# 복습

In [1]:
import pandas as pd

In [2]:
from pandas import Series, DataFrame

In [3]:
import numpy as np
np.random.seed(12345)
import matplotlib.pyplot as plt
plt.rc('figure', figsize=(10, 6))
PREVIOUS_MAX_ROWS = pd.options.display.max_rows
pd.options.display.max_rows = 20
np.set_printoptions(precision=4, suppress=True)

### 1. 데이터 로딩: pd.read_csv()

 'datasets/exercise/insurance.csv'파일 읽어서 df에 할당하기

In [4]:
df = pd.read_csv('insurance.csv')
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

### 2. df의 상위 5개 항목 출력

In [7]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

### 3. df의 요약(평균, 4분위수, 전체 갯수 등)

In [8]:
df.describe()             # 수치형 데이터들만 나온다ㅎㅎ

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

### 4. 신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기

In [9]:
df['double bmi'] = df['bmi']*2
df

age     sex     bmi  children smoker     region      charges  double bmi
0      19  female  27.900         0    yes  southwest  16884.92400       55.80
1      18    male  33.770         1     no  southeast   1725.55230       67.54
2      28    male  33.000         3     no  southeast   4449.46200       66.00
3      33    male  22.705         0     no  northwest  21984.47061       45.41
4      32    male  28.880         0     no  northwest   3866.85520       57.76
...   ...     ...     ...       ...    ...        ...          ...         ...
1333   50    male  30.970         3     no  northwest  10600.54830       61.94
1334   18  female  31.920         0     no  northeast   2205.98080       63.84
1335   18  female  36.850         0     no  southeast   1629.83350       73.70
1336   21  female  25.800         0     no  southwest   2007.94500       51.60
1337   61  female  29.070         0    yes  northwest  29141.36030       58.14

[1338 rows x 8 columns]

### 5. 신규 컬럼(debt)을 생성하고, 0으로 채우기

In [10]:
df['debt'] = 0
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     0  
2          66.00     0  
3          45.41     0  
4          57.76     0  
...          ...   ...  
1333       61.94     0  
1334       63.84     0  
1335       73.70     0  
1336       51.60     0  
1337       58.14     0  

[1338 rows x 9 columns]

### 6. debt에 0부터 값이 증가하도록 채우기

In [11]:
series = np.arange(0,1338,1)            # df['debt'] = np.arange(1338) 이렇게도 가능
df['debt']=series                       # df['debt'] = np.arange(0,len(df)) 요렇게도~
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41     3  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]

### 7. 1번 인덱스의 정보 출력:iloc

In [12]:
df.iloc[1]        # iloc 는 인덱스의 정보를 출력한다

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double bmi        67.54
debt                  1
Name: 1, dtype: object

### 8. 3,5,8번 인덱스의 'debt'값을 12, 13, 14로 설정하기

In [13]:
df.loc[[3], ['debt']] = 12        # val = pd.Series([12,13,14], index=[3,5,8])
df.loc[[5], ['debt']] = 13        # df['debt'] = val
df.loc[[8], ['debt']] = 14        # df                      이렇게도 하더라...

df                               # df['debt'][3] =12        이렇게도 할 수 있다 ㅎㅎ         
                                 # df.loc[3,'debt'] =12   

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41    12  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]

### 9. 'double bmi'열 삭제하기

In [15]:
del df["double bmi"]
df

age     sex     bmi  children smoker     region      charges  debt
0      19  female  27.900         0    yes  southwest  16884.92400     0
1      18    male  33.770         1     no  southeast   1725.55230     1
2      28    male  33.000         3     no  southeast   4449.46200     2
3      33    male  22.705         0     no  northwest  21984.47061    12
4      32    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  1333
1334   18  female  31.920         0     no  northeast   2205.98080  1334
1335   18  female  36.850         0     no  southeast   1629.83350  1335
1336   21  female  25.800         0     no  southwest   2007.94500  1336
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

### 10. 'age'열, 'sex'열만 출력하기

In [16]:
df[['age', 'sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

### 11. 함수 생성, 적용(apply/map)

나이를 1살씩 증가시키기

In [5]:
f = lambda x: x+1
#def plus_one (x):
#    x=x+1
#    return x

df['age'].apply(f)


#df['age'].apply(x=x+1)

0       20
1       19
2       29
3       34
4       33
        ..
1333    51
1334    19
1335    19
1336    22
1337    62
Name: age, Length: 1338, dtype: int64

### 12. 함수 생성, 적용(apply/map)

charges의 표시형식을 소수점 아래 2자리로 바꾸기

In [6]:
ff = lambda x: '%.2f' % x
df['charges'].apply(ff)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: object

### 13. 함수 생성, 적용(applymap)

데이터프레임 전체 값을 대괄호[ ] 로 감싸기

예: 	[19]	[female]	[27.9]	[0]	[yes]	[southwest]	[16884.924]

In [11]:
def c (a):
    a=[a]
    return a
df.applymap(c)           # apply map는 데이터 프레임 원소 전체에 적용

# ff = lambda x: [x]
# df.applympap(ff)

age       sex       bmi children smoker       region        charges
0     [19]  [female]    [27.9]      [0]  [yes]  [southwest]    [16884.924]
1     [18]    [male]   [33.77]      [1]   [no]  [southeast]    [1725.5523]
2     [28]    [male]    [33.0]      [3]   [no]  [southeast]     [4449.462]
3     [33]    [male]  [22.705]      [0]   [no]  [northwest]  [21984.47061]
4     [32]    [male]   [28.88]      [0]   [no]  [northwest]    [3866.8552]
...    ...       ...       ...      ...    ...          ...            ...
1333  [50]    [male]   [30.97]      [3]   [no]  [northwest]   [10600.5483]
1334  [18]  [female]   [31.92]      [0]   [no]  [northeast]    [2205.9808]
1335  [18]  [female]   [36.85]      [0]   [no]  [southeast]    [1629.8335]
1336  [21]  [female]    [25.8]      [0]   [no]  [southwest]     [2007.945]
1337  [61]  [female]   [29.07]      [0]  [yes]  [northwest]   [29141.3603]

[1338 rows x 7 columns]

### 14. 컬럼 정렬(알파벳 순서대로) : sort_index
age, bmi, charges... 순으로 정렬

In [17]:
df.sort_index(axis=1)

age     bmi      charges  children  debt     region     sex smoker
0      19  27.900  16884.92400         0     0  southwest  female    yes
1      18  33.770   1725.55230         1     1  southeast    male     no
2      28  33.000   4449.46200         3     2  southeast    male     no
3      33  22.705  21984.47061         0    12  northwest    male     no
4      32  28.880   3866.85520         0     4  northwest    male     no
...   ...     ...          ...       ...   ...        ...     ...    ...
1333   50  30.970  10600.54830         3  1333  northwest    male     no
1334   18  31.920   2205.98080         0  1334  northeast  female     no
1335   18  36.850   1629.83350         0  1335  southeast  female     no
1336   21  25.800   2007.94500         0  1336  southwest  female     no
1337   61  29.070  29141.36030         0  1337  northwest  female    yes

[1338 rows x 8 columns]

### 15. 값 정렬(age) : sort_values
나이 순으로 정렬

In [14]:
df['age'].sort_values()
#df['age'].sort_values(by='age')

1248    18
482     18
492     18
525     18
529     18
        ..
398     64
335     64
378     64
1265    64
635     64
Name: age, Length: 1338, dtype: int64

### 16. 평균구하기: mean

In [19]:
df.mean()

age            39.207025
bmi            30.663397
children        1.094918
charges     13270.422265
debt          668.517190
dtype: float64

### 17. 공분산 구하기: cov

In [20]:
df.cov

<bound method DataFrame.cov of       age     sex     bmi  children smoker     region      charges  debt
0      19  female  27.900         0    yes  southwest  16884.92400     0
1      18    male  33.770         1     no  southeast   1725.55230     1
2      28    male  33.000         3     no  southeast   4449.46200     2
3      33    male  22.705         0     no  northwest  21984.47061    12
4      32    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  1333
1334   18  female  31.920         0     no  northeast   2205.98080  1334
1335   18  female  36.850         0     no  southeast   1629.83350  1335
1336   21  female  25.800         0     no  southwest   2007.94500  1336
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]>

### 18. 남자, 여자가 각각 몇명있는지 구하기

In [21]:
df["sex"].value_counts()         

male      676
female    662
Name: sex, dtype: int64

In [15]:
df['sex'].unique()        # 유일 값만 출력 

array(['female', 'male'], dtype=object)

### 19. 자녀(children)이 3명인 데이터 출력하기

In [22]:
df[df['children']==3]       

age     sex     bmi  children smoker     region      charges  debt
2      28    male  33.000         3     no  southeast   4449.46200     2
7      37  female  27.740         3     no  northwest   7281.50560     7
25     59  female  27.720         3     no  southeast  14001.13380    25
36     62  female  32.965         3     no  northwest  15612.19335    36
54     40  female  28.690         3     no  northwest   8059.67910    54
...   ...     ...     ...       ...    ...        ...          ...   ...
1301   62    male  30.875         3    yes  northwest  46718.16325  1301
1314   30  female  23.655         3    yes  northwest  18765.87545  1314
1320   31    male  31.065         3     no  northwest   5425.02335  1320
1332   52  female  44.700         3     no  southwest  11411.68500  1332
1333   50    male  30.970         3     no  northwest  10600.54830  1333

[157 rows x 8 columns]

### 20. 자녀(children)가 5명 이상인 데이터 출력하기

In [23]:
df[df['children']>=5]

age     sex     bmi  children smoker     region      charges  debt
32     19  female  28.600         5     no  southwest   4687.79700    32
71     31    male  28.500         5     no  northeast   6799.45800    71
166    20  female  37.000         5     no  southwest   4830.63000   166
413    25    male  23.900         5     no  southwest   5080.09600   413
425    45    male  24.310         5     no  southeast   9788.86590   425
438    52  female  46.750         5     no  southeast  12592.53450   438
568    49  female  31.900         5     no  southwest  11552.90400   568
640    33    male  42.400         5     no  southwest   6666.24300   640
877    33    male  33.440         5     no  southeast   6653.78860   877
932    46    male  25.800         5     no  southwest  10096.97000   932
937    39  female  24.225         5     no  northwest   8965.79575   937
969    39  female  34.320         5     no  southeast   8596.82780   969
984    20    male  30.115         5     no  northeast   4915.05985   984
1085   39  female  18.300         5    yes  southwest  19023.26000  1085
1116   41    male  29.640         5     no  northeast   9222.40260  1116
1130   39  female  23.870         5     no  southeast   8582.30230  1130
1245   28    male  24.300         5     no  southwest   5615.36900  1245
1272   43    male  25.520         5     no  southeast  14478.33015  1272

### 21. bmi가  17 미만인 데이터 출력하기

In [24]:
df[df['bmi']<17]

age     sex     bmi  children smoker     region     charges  debt
172    18    male  15.960         0     no  northeast  1694.79640   172
428    21  female  16.815         1     no  northeast  3167.45585   428
1226   38    male  16.815         2     no  northeast  6640.54485  1226

### 22. debt가 null이 아닌데이터 출력하기

In [25]:
df[df['debt']!='null']                   #df[df['debt'].notnull()]

age     sex     bmi  children smoker     region      charges  debt
0      19  female  27.900         0    yes  southwest  16884.92400     0
1      18    male  33.770         1     no  southeast   1725.55230     1
2      28    male  33.000         3     no  southeast   4449.46200     2
3      33    male  22.705         0     no  northwest  21984.47061    12
4      32    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  1333
1334   18  female  31.920         0     no  northeast   2205.98080  1334
1335   18  female  36.850         0     no  southeast   1629.83350  1335
1336   21  female  25.800         0     no  southwest   2007.94500  1336
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

### 23. 'age'가 60보다 큰 값만 출력하기

In [26]:
df[df['age']>60]

age     sex     bmi  children smoker     region      charges  debt
11     62  female  26.290         0    yes  southeast  27808.72510    11
26     63  female  23.085         0     no  northeast  14451.83515    26
33     63    male  28.310         0     no  northwest  13770.09790    33
36     62  female  32.965         3     no  northwest  15612.19335    36
62     64    male  24.700         1     no  northwest  30166.61817    62
...   ...     ...     ...       ...    ...        ...          ...   ...
1301   62    male  30.875         3    yes  northwest  46718.16325  1301
1321   62    male  26.695         0    yes  northeast  28101.33305  1321
1322   62    male  38.830         0     no  southeast  12981.34570  1322
1325   61    male  33.535         0     no  northeast  13143.33665  1325
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[91 rows x 8 columns]

In [3]:
import pandas as pd
import numpy as np

In [5]:
df = pd.DataFrame(np.random.randn(6,3))
df

0         1         2
0  0.566301 -1.207405  0.105106
1  0.644516  0.949858 -0.478021
2  1.754833 -0.190788 -0.750726
3  1.387831  0.328768 -0.006782
4 -0.143848 -0.068559 -0.653541
5  2.261548  0.309318 -1.195534

In [7]:
df.iloc[1:,1]

1    0.949858
2   -0.190788
3    0.328768
4   -0.068559
5    0.309318
Name: 1, dtype: float64